bin/fuel-core/src/cli/snapshot.rs

```diff
@@ -118,31 +154,59 @@ pub fn exec(command: Command) -> anyhow::Result<()> {
             let encoding = encoding_command
                 .map(|f| f.encoding())
                 .unwrap_or_else(|| Encoding::Json);
-            full_snapshot(chain_config, &output_dir, encoding, &db)
+            full_snapshot(chain_config, &output_dir, encoding, db)
         }
         SubCommands::Contract { contract_id } => {
-            contract_snapshot(&db, contract_id, &output_dir)
+            contract_snapshot(db, contract_id, &output_dir)
         }
     }
 }
 
 fn contract_snapshot(
-    db: impl ChainStateDb,
+    db: CombinedDatabase,
     contract_id: ContractId,
     output_dir: &Path,
 ) -> Result<(), anyhow::Error> {
     std::fs::create_dir_all(output_dir)?;
 
-    let (contract, state, balance) = db.get_contract_by_id(contract_id)?;
-    let block = db.get_last_block()?;
-
-    let metadata = write_metadata(output_dir, Encoding::Json)?;
-    let mut writer = StateWriter::for_snapshot(&metadata)?;
-
-    writer.write_contracts(vec![contract])?;
-    writer.write_contract_state(state)?;
-    writer.write_contract_balance(balance)?;
+    let code = db
+        .on_chain()
+        .entries::<ContractsRawCode>(Some(contract_id.as_ref()), IterDirection::Forward)
+        .next()
+        .ok_or_else(|| {
+            anyhow::anyhow!("contract code not found! id: {:?}", contract_id)
+        })??;
+
+    let utxo = db
+        .on_chain()
+        .entries::<ContractsLatestUtxo>(
+            Some(contract_id.as_ref()),
+            IterDirection::Forward,
+        )
+        .next()
+        .ok_or_else(|| {
+            anyhow::anyhow!("contract utxo not found! id: {:?}", contract_id)
+        })??;
+
+    let state = db
+        .on_chain()
+        .entries::<ContractsState>(Some(contract_id.as_ref()), IterDirection::Forward)
+        .try_collect()?;
+
+    let balance = db
+        .on_chain()
+        .entries::<ContractsAssets>(Some(contract_id.as_ref()), IterDirection::Forward)
+        .try_collect()?;
+
+    let block = db.on_chain().latest_block()?;
+    let mut writer = SnapshotWriter::json(output_dir);
+
+    writer.write(vec![code])?;
+    writer.write(vec![utxo])?;
+    writer.write(state)?;
+    writer.write(balance)?;
     writer.write_block_data(*block.header().height(), block.header().da_height)?;
+    writer.write_chain_config(&ChainConfig::local_testnet())?;
```

Comment:
A snapshot must contain the chain info. 

When reading a snapshot some things are eagerly read -- like the chain config and block heights -- so `reader.chain_config()` don't have to have to return a `Result`.

So even though we may not need the chain config in the e2e tests, we still write it because of the eager-reading of it by the `SnapshotReader`.